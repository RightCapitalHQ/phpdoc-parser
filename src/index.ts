export { Lexer } from './phpdoc-parser/lexer/lexer';
export { ConstExprParser } from './phpdoc-parser/parser/const-expr-parser';
export { StringUnescaper } from './phpdoc-parser/parser/string-unescaper';
export { ParserException } from './phpdoc-parser/parser/parser-exception';
export { TokenIterator } from './phpdoc-parser/parser/token-iterator';
export { PhpDocParser } from './phpdoc-parser/parser/php-doc-parser';
export { TypeParser } from './phpdoc-parser/parser/type-parser';
export { BaseNode } from './phpdoc-parser/ast/base-node';
export type { Node } from './phpdoc-parser/ast/node';
export { NodeTraverser } from './phpdoc-parser/ast/node-traverser';
export { CloningVisitor } from './phpdoc-parser/ast/node-visitor/cloning-visitor';
export type { NodeVisitor } from './phpdoc-parser/ast/node-visitor';
export { NodeTraverserState, Attribute } from './phpdoc-parser/ast/types';
export { IdentifierTypeNode } from './phpdoc-parser/ast/type/identifier-type-node';
export { CallableTypeParameterNode } from './phpdoc-parser/ast/type/callable-type-parameter-node';
export { OffsetAccessTypeNode } from './phpdoc-parser/ast/type/offset-access-type-node';
export { ArrayShapeItemNode } from './phpdoc-parser/ast/type/array-shape-item-node';
export { ConditionalTypeNode } from './phpdoc-parser/ast/type/conditional-type-node';
export { ConditionalTypeForParameterNode } from './phpdoc-parser/ast/type/conditional-type-for-parameter-node';
export { ThisTypeNode } from './phpdoc-parser/ast/type/this-type-node';
export { IntersectionTypeNode } from './phpdoc-parser/ast/type/intersection-type-node';
export { ObjectShapeItemNode } from './phpdoc-parser/ast/type/object-shape-item-node';
export { UnionTypeNode } from './phpdoc-parser/ast/type/union-type-node';
export { TypeNode } from './phpdoc-parser/ast/type/type-node';
export { GenericTypeNode } from './phpdoc-parser/ast/type/generic-type-node';
export { CallableTypeNode } from './phpdoc-parser/ast/type/callable-type-node';
export { ArrayTypeNode } from './phpdoc-parser/ast/type/array-type-node';
export { ArrayShapeNode } from './phpdoc-parser/ast/type/array-shape-node';
export { InvalidTypeNode } from './phpdoc-parser/ast/type/invalid-type-node';
export { ConstTypeNode } from './phpdoc-parser/ast/type/const-type-node';
export { ObjectShapeNode } from './phpdoc-parser/ast/type/object-shape-node';
export { NullableTypeNode } from './phpdoc-parser/ast/type/nullable-type-node';
export { AbstractNodeVisitor } from './phpdoc-parser/ast/abstract-node-visitor';
export { ConstExprFloatNode } from './phpdoc-parser/ast/const-expr/const-expr-float-node';
export { ConstExprArrayItemNode } from './phpdoc-parser/ast/const-expr/const-expr-array-item-node';
export { ConstExprFalseNode } from './phpdoc-parser/ast/const-expr/const-expr-false-node';
export { QuoteAwareConstExprStringNode } from './phpdoc-parser/ast/const-expr/quote-aware-const-expr-string-node';
export { ConstExprNullNode } from './phpdoc-parser/ast/const-expr/const-expr-null-node';
export { ConstExprIntegerNode } from './phpdoc-parser/ast/const-expr/const-expr-integer-node';
export { ConstExprNode } from './phpdoc-parser/ast/const-expr/const-expr-node';
export { ConstExprArrayNode } from './phpdoc-parser/ast/const-expr/const-expr-array-node';
export { ConstExprStringNode } from './phpdoc-parser/ast/const-expr/const-expr-string-node';
export { ConstExprTrueNode } from './phpdoc-parser/ast/const-expr/const-expr-true-node';
export { ConstFetchNode } from './phpdoc-parser/ast/const-expr/const-fetch-node';
export { ParamOutTagValueNode } from './phpdoc-parser/ast/php-doc/param-out-tag-value-node';
export { PhpDocNode } from './phpdoc-parser/ast/php-doc/php-doc-node';
export { AssertTagValueNode } from './phpdoc-parser/ast/php-doc/assert-tag-value-node';
export { UsesTagValueNode } from './phpdoc-parser/ast/php-doc/uses-tag-value-node';
export { InvalidTagValueNode } from './phpdoc-parser/ast/php-doc/invalid-tag-value-node';
export { MethodTagValueNode } from './phpdoc-parser/ast/php-doc/method-tag-value-node';
export { DeprecatedTagValueNode } from './phpdoc-parser/ast/php-doc/deprecated-tag-value-node';
export { ExtendsTagValueNode } from './phpdoc-parser/ast/php-doc/extends-tag-value-node';
export { ParamTagValueNode } from './phpdoc-parser/ast/php-doc/param-tag-value-node';
export { GenericTagValueNode } from './phpdoc-parser/ast/php-doc/generic-tag-value-node';
export { TypeAliasTagValueNode } from './phpdoc-parser/ast/php-doc/type-alias-tag-value-node';
export { AssertTagPropertyValueNode } from './phpdoc-parser/ast/php-doc/assert-tag-property-value-node';
export { TypelessParamTagValueNode } from './phpdoc-parser/ast/php-doc/typeless-param-tag-value-node';
export { VarTagValueNode } from './phpdoc-parser/ast/php-doc/var-tag-value-node';
export { PhpDocTextNode } from './phpdoc-parser/ast/php-doc/php-doc-text-node';
export { MethodTagValueParameterNode } from './phpdoc-parser/ast/php-doc/method-tag-value-parameter-node';
export { AssertTagMethodValueNode } from './phpdoc-parser/ast/php-doc/assert-tag-method-value-node';
export { TemplateTagValueNode } from './phpdoc-parser/ast/php-doc/template-tag-value-node';
export { ImplementsTagValueNode } from './phpdoc-parser/ast/php-doc/implements-tag-value-node';
export { ThrowsTagValueNode } from './phpdoc-parser/ast/php-doc/throws-tag-value-node';
export { SelfOutTagValueNode } from './phpdoc-parser/ast/php-doc/self-out-tag-value-node';
export { PropertyTagValueNode } from './phpdoc-parser/ast/php-doc/property-tag-value-node';
export { TypeAliasImportTagValueNode } from './phpdoc-parser/ast/php-doc/type-alias-import-tag-value-node';
export { PhpDocChildNode } from './phpdoc-parser/ast/php-doc/php-doc-child-node';
export { MixinTagValueNode } from './phpdoc-parser/ast/php-doc/mixin-tag-value-node';
export { PhpDocTagNode } from './phpdoc-parser/ast/php-doc/php-doc-tag-node';
export { PhpDocTagValueNode } from './phpdoc-parser/ast/php-doc/php-doc-tag-value-node';
export { ReturnTagValueNode } from './phpdoc-parser/ast/php-doc/return-tag-value-node';
export { Printer } from './phpdoc-parser/printer/printer';
export { PhpDocTypeNodeToTypescriptTypeNodeTranspiler } from './phpdoc-parser/transpiler/php-doc-to-typescript-type-transpiler';
export * from './phpdoc-parser/ast/php-doc/helpers';
export * from './phpdoc-parser/transpiler/helpers';
