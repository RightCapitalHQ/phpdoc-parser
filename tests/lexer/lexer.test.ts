import { it, expect } from 'vitest';
import { Lexer } from '../../src';

// Comes from Laravel https://github.com/laravel/framework/blob/10.x/src/Illuminate/Collections/Collection.php
const commonSource = `
/**
 * @template TKey of array-key
 *
 * @template-covariant TValue
 *
 * @implements \\ArrayAccess<TKey, TValue>
 * @implements \\Illuminate\\Support\\Enumerable<TKey, TValue>
 */
`.trim();

it('test tokenize for a very common doc annotation', () => {
  const lexer = new Lexer();
  const tokens = lexer.tokenize(commonSource);
  expect(tokens).toMatchSnapshot('commonSource');
});
