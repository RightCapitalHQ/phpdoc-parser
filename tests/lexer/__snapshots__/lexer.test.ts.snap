// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test tokenize for a very common doc annotation: commonSource 1`] = `
[
  [
    "/**",
    "TOKEN_OPEN_PHPDOC",
    1,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    1,
  ],
  [
    "@property",
    "TOKEN_PHPDOC_TAG",
    2,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    2,
  ],
  [
    "float",
    "TOKEN_IDENTIFIER",
    2,
  ],
  [
    "|",
    "TOKEN_UNION",
    2,
  ],
  [
    "null",
    "TOKEN_IDENTIFIER",
    2,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    2,
  ],
  [
    "$cash_balance",
    "TOKEN_VARIABLE",
    2,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    2,
  ],
  [
    "@property-read",
    "TOKEN_PHPDOC_TAG",
    3,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    3,
  ],
  [
    "string",
    "TOKEN_IDENTIFIER",
    3,
  ],
  [
    "|",
    "TOKEN_UNION",
    3,
  ],
  [
    "null",
    "TOKEN_IDENTIFIER",
    3,
  ],
  [
    "         ",
    "TOKEN_HORIZONTAL_WS",
    3,
  ],
  [
    "$default_name",
    "TOKEN_VARIABLE",
    3,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    3,
  ],
  [
    "@property-read",
    "TOKEN_PHPDOC_TAG",
    4,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    4,
  ],
  [
    "int",
    "TOKEN_IDENTIFIER",
    4,
  ],
  [
    "|",
    "TOKEN_UNION",
    4,
  ],
  [
    "null",
    "TOKEN_IDENTIFIER",
    4,
  ],
  [
    "            ",
    "TOKEN_HORIZONTAL_WS",
    4,
  ],
  [
    "$refresh_code",
    "TOKEN_VARIABLE",
    4,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    4,
  ],
  [
    "@property-read",
    "TOKEN_PHPDOC_TAG",
    5,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    5,
  ],
  [
    "\\Carbon\\Carbon",
    "TOKEN_IDENTIFIER",
    5,
  ],
  [
    "|",
    "TOKEN_UNION",
    5,
  ],
  [
    "null",
    "TOKEN_IDENTIFIER",
    5,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    5,
  ],
  [
    "$refreshed_at",
    "TOKEN_VARIABLE",
    5,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    5,
  ],
  [
    "@property-read",
    "TOKEN_PHPDOC_TAG",
    6,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    6,
  ],
  [
    "string",
    "TOKEN_IDENTIFIER",
    6,
  ],
  [
    "|",
    "TOKEN_UNION",
    6,
  ],
  [
    "null",
    "TOKEN_IDENTIFIER",
    6,
  ],
  [
    "         ",
    "TOKEN_HORIZONTAL_WS",
    6,
  ],
  [
    "$refreshed_at_friendly",
    "TOKEN_VARIABLE",
    6,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    6,
  ],
  [
    "@property-read",
    "TOKEN_PHPDOC_TAG",
    7,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    7,
  ],
  [
    "int",
    "TOKEN_IDENTIFIER",
    7,
  ],
  [
    "                 ",
    "TOKEN_HORIZONTAL_WS",
    7,
  ],
  [
    "$signed_balance",
    "TOKEN_VARIABLE",
    7,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    7,
  ],
  [
    "@property-read",
    "TOKEN_PHPDOC_TAG",
    8,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    8,
  ],
  [
    "int",
    "TOKEN_IDENTIFIER",
    8,
  ],
  [
    "|",
    "TOKEN_UNION",
    8,
  ],
  [
    "null",
    "TOKEN_IDENTIFIER",
    8,
  ],
  [
    "            ",
    "TOKEN_HORIZONTAL_WS",
    8,
  ],
  [
    "$tax_advantage_id",
    "TOKEN_VARIABLE",
    8,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    8,
  ],
  [
    "@property-read",
    "TOKEN_PHPDOC_TAG",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "float",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    "               ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "$total_cost_basis",
    "TOKEN_VARIABLE",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "total",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "cost",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "to",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "buy",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    ",",
    "TOKEN_COMMA",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "both",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "stocks",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "and",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    9,
  ],
  [
    "funds",
    "TOKEN_IDENTIFIER",
    9,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    9,
  ],
  [
    "@property-read",
    "TOKEN_PHPDOC_TAG",
    10,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "float",
    "TOKEN_IDENTIFIER",
    10,
  ],
  [
    "               ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "$total_value",
    "TOKEN_VARIABLE",
    10,
  ],
  [
    "      ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "total",
    "TOKEN_IDENTIFIER",
    10,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "market",
    "TOKEN_IDENTIFIER",
    10,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "value",
    "TOKEN_IDENTIFIER",
    10,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "+",
    "TOKEN_OTHER",
    10,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "signed",
    "TOKEN_IDENTIFIER",
    10,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "cash",
    "TOKEN_IDENTIFIER",
    10,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    10,
  ],
  [
    "balance",
    "TOKEN_IDENTIFIER",
    10,
  ],
  [
    "
 *",
    "TOKEN_PHPDOC_EOL",
    10,
  ],
  [
    "
 * ",
    "TOKEN_PHPDOC_EOL",
    11,
  ],
  [
    "@method",
    "TOKEN_PHPDOC_TAG",
    12,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    12,
  ],
  [
    "static",
    "TOKEN_IDENTIFIER",
    12,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    12,
  ],
  [
    "\\Illuminate\\Database\\Eloquent\\Builder",
    "TOKEN_IDENTIFIER",
    12,
  ],
  [
    " ",
    "TOKEN_HORIZONTAL_WS",
    12,
  ],
  [
    "included",
    "TOKEN_IDENTIFIER",
    12,
  ],
  [
    "(",
    "TOKEN_OPEN_PARENTHESES",
    12,
  ],
  [
    ")",
    "TOKEN_CLOSE_PARENTHESES",
    12,
  ],
  [
    "
 ",
    "TOKEN_PHPDOC_EOL",
    12,
  ],
  [
    "*/",
    "TOKEN_CLOSE_PHPDOC",
    13,
  ],
  [
    "",
    "TOKEN_END",
    13,
  ],
]
`;
